src/LifecycleHooks.lua

```diff
@@ -46,119 +46,60 @@ end
-			return
-		end
-
-		back[type] = popped[type]
-	end
-
-	pushHooksUp(TestEnum.NodeType.BeforeAll)
-	pushHooksUp(TestEnum.NodeType.AfterAll)
 end
 
 function LifecycleHooks:pushHooksFrom(planNode)
 	assert(planNode ~= nil)
 
 	table.insert(self._stack, {
-		[TestEnum.NodeType.BeforeAll] = self:_getBeforeAllHooksUncalledAtCurrentLevel(planNode.children),
-		[TestEnum.NodeType.AfterAll] = self:_getAfterAllHooksUncalledAtCurrentLevel(planNode.children),
+		[TestEnum.NodeType.BeforeAll] = self:_getHooksOfType(planNode.children, TestEnum.NodeType.BeforeAll),
+		[TestEnum.NodeType.AfterAll] = self:_getHooksOfType(planNode.children, TestEnum.NodeType.AfterAll),
 		[TestEnum.NodeType.BeforeEach] = self:_getHooksOfType(planNode.children, TestEnum.NodeType.BeforeEach),
 		[TestEnum.NodeType.AfterEach] = self:_getHooksOfType(planNode.children, TestEnum.NodeType.AfterEach),
 	})
 end
 
-function LifecycleHooks:getPendingBeforeAllHooks()
-	return self:_getAndClearPendingHooks(TestEnum.NodeType.BeforeAll)
-end
-
-function LifecycleHooks:getAfterAllHooks()
-	if #self._stack > 0 then
-		return self:_getAndClearPendingHooks(TestEnum.NodeType.AfterAll)
-	else
-		return {}
-	end
-end
-
 --[[
-	Return any hooks that have not yet been returned for this key and clear those hooks
+	Get all currently uncalled beforeAll hooks, and remove them from the stack.
 ]]
-function LifecycleHooks:_getAndClearPendingHooks(key)
-	assert(key ~= nil)
-
-	if #self._stack > 0 then
-
-		local back = self._stack[#self._stack]
-
-		local hooks = back[key]
-
-		back[key] = {}
-
-		return hooks
-	else
-		return {}
-	end
-end
-
---[[
-	Transfers uncalled beforeAll and afterAll hooks down the stack
-]]
-function LifecycleHooks:_getBeforeAllHooksUncalledAtCurrentLevel(childNodes)
-	local hookType = TestEnum.NodeType.BeforeAll
-	local hooks = self:_getHooksOfTypeFromBackOfStack(hookType)
-
-	for _, hook in pairs(self:_getHooksOfType(childNodes, hookType)) do
-		table.insert(hooks, hook)
-	end
-
-	return hooks
-end
-
-function LifecycleHooks:_getAfterAllHooksUncalledAtCurrentLevel(childNodes)
-	local hookType = TestEnum.NodeType.AfterAll
-	local hooks = self:_getHooksOfTypeFromBackOfStack(hookType)
+function LifecycleHooks:getPendingBeforeAllHooks()
+	local key = TestEnum.NodeType.BeforeAll
+	local hooks = {}
 
-	for _, hook in pairs(self:_getHooksOfType(childNodes, hookType)) do
-		table.insert(hooks, 1, hook)
+	for _, level in ipairs(self._stack) do
+		for _, hook in ipairs(level[key]) do
+			table.insert(hooks, hook)
+		end
+		level[key] = {}
 	end
 
 	return hooks
 end
 
-function LifecycleHooks:_getHooksOfTypeFromBackOfStack(hookType)
-	assert(hookType, "Expected hookType to be an argument")
-
-	local currentBack = self:_getBackOfStack()
-
+--[[
+	Get all uncalled afterAll hooks from the back of the stack and remove them.
+]]
+function LifecycleHooks:getAfterAllHooks()
+	local key = TestEnum.NodeType.AfterAll
 	local hooks = {}
 
+	local currentBack = self._stack[#self._stack] or nil
```

Comment:
This shouldn't need an `or nil` -- it'll be nil already if the stack is empty.

```suggestion
	local currentBack = self._stack[#self._stack]
```